### ISSUE TITLE
Replace (#) with (&)

### ISSUE THREAD

**Issue by RyanGlScott** [ku-fpg/blank-canvas#66]
Right now, we have a function `(#) :: a -> (a -> b) -> b` such that `a # f = f a`. It works, but there's been talk of using the `(#)` as a natural transformation instead (currently in the [`wakarusa`](https://github.com/ku-fpg/wakarusa/blob/5f1d893cdb11c010b41c7d9b46ab9c9bfd3da00b/src/Control/Natural.hs#L11-14) repo). With this, we could define natural transformations such as:

```haskell
instance Transformation Canvas IO DeviceContext where
    (#) = send
```

However, this would not work for things that we currently use the `blank-canvas` `(#)` for, such as `grd # addColorStop(0, "#8ED6FF")` and `(defFont [sansSerif]) { fontSize = 30 # pt }`. I think the best way to mitigate such a change would be to use the [`(&) :: a -> (a -> b) -> b` function](http://hackage.haskell.org/package/base-4.8.0.0/candidate/docs/Data-Function.html#v:-38-) that was introduced in GHC 7.10 in `Data.Function` (if one is using an older version of GHC, it's easy to redefine it: `a & f = f a`). There's also a bit of history in that `(&)` is [also used in `lens`](http://hackage.haskell.org/package/lens-4.7/docs/Control-Lens-Operators.html#v:-38-), so it's not a completely arbitrary choice of name.

@andygill, what are your thoughts?

**Comment by andygill** [comment]
Idea: make `grd` a natural Transformation, and then `grd # addColorStop(0, "#8ED6FF")` can be made to work, with a bit of persuasion.

**Comment by RyanGlScott** [comment]
How would a `CanvasGradient` instance for `Natural` work? I ask since the type signature for `addColorStop`—`(Double, Text) -> CanvasGradient -> Canvas ()`—is entirely monomorphic, so I don't see how it admits an instance of `Natural`, which requires two parametrically polymorphic types.

**Comment by RyanGlScott** [comment]
The closest thing I can think of is

```haskell
instance Natural ((->) CanvasGradient) Identity CanvasGradient where
    grd # f = Identity (f grd)
```

but that would make `grd # addColorStop(0, "#8ED6FF")` return something of type `Identity (Canvas ())`, not `Canvas ()`.

**Comment by andygill** [comment]
Currently

     addColorStop :: CanvasColor color => (Interval, color) -> CanvasGradient -> Canvas ()

Here's the trick

    addColorStop :: CanvasColor color => (Interval, color) -> ColorCommand ()
    type CanvasGradient = Natural ColorCommand Canvas

Thus

    grd # addColorStop(0, "#8ED6FF")

**Comment by RyanGlScott** [comment]
I'm a bit confused. What is `ColorCommand`?

**Comment by andygill** [comment]
````.haskell
data ColorCommand where
  AddColorStop' :: CanvasColor color => (Interval, color) -> ColorCommand ()
addColorStop = AddColorStop'
````

Which means what is grd?

````.haskell
grd = Natural $ \ f -> case f of 
  AddColorStop' (inv,col) -> Command (AddColorStop (off,rep) <<original-grd :: CanvasGradient>>)
```
So the `grd` "adds" the original CanvasGradient, when the Natural transformation is applied.

**Comment by RyanGlScott** [comment]
Ah. So if I'm understanding you correctly, we would (1) create a separate ADT for "member functions" of `CanvasGradient`, (2) change `CanvasGradient` from a wrapper around a buffer number to a `Natural`, (3) change `createLinearGradient` and `createRadialGradient` to return a `Natural` that gives you `original_grd` and (4) change the type signature of `addColorStop` to be compatible with this scheme.

This sounds pretty promising, but it would break some existing `blank-canvas` code in the wild right now (e.g., [`diagrams-canvas` currently expects `addColorStop` to take `grd` as an argument](https://github.com/diagrams/diagrams-canvas/blob/6e4836dec5ba897fa3db333eb31f7ec262873ca1/src/Diagrams/Backend/Canvas.hs#L262)).

Also, I'm curious to know if there's a way to compose `ColorCommands` in some way. I'm thinking perhaps an "effectful" version `(#)`, perhaps using a naming convention inspired by `lens`:

```haskell
(<#) :: (Transformation f g t, Applicative g) => t -> f a -> g t
t <# f = (t # f) *> pure t
```

This way, you could chain `addColorStop`s like this:

```haskell
grd <# addColorStop(0.0, "red") <# addColorStop(0.5, "green") # addColorStop(1.0, "blue")
```

**Comment by andygill** [comment]
How about

````.haskell
grd # do addColorStop(0.0, "red")
         addColorStop(0.5, "green")
         addColorStop(1.0, "blue")

This is possible with some trickery.

**Comment by RyanGlScott** [comment]
You're right. I suppose we could use a similar trick to what `Canvas` does (by having `Bind` and `Return` constructors), or we could wrap it in a `Free` monad.
